File: archives/non_decreasing_array_665/Solution.java
package non_decreasing_array_665;

/**
 * Created by Zhiyong Pan on 2021-02-03.
 */
class Solution {
    public boolean checkPossibility(int[] nums) {
        int modified = 0;
        for (int i = 0; i < nums.length - 1; ++i) {
            if (nums[i] > nums[i + 1]) {
                if (i > 0) {
                    if (nums[i - 1] <= nums[i + 1]) {
                        // Our best option is lowering [i] to [i - 1].
                        nums[i] = nums[i - 1];
                    } else {
                        // Our only option is raising [i+1] to [i].
                        nums[i + 1] = nums[i];
                    }
                } else {
                    // Our best option is lowering [i].
                    nums[i] = nums[i + 1];
                }

                if (++modified > 1)
                    return false;
            }
        }
        return true;
    }
}

File: archives/non_decreasing_array_665/Solution.cpp
/* 665. Non-decreasing Array
 * https://leetcode.com/problems/non-decreasing-array/
 * --
 * Zhiyong 2021-05-04
 */
class Solution {
public:
    bool checkPossibility(vector<int>& nums) {
        bool modified = false;
        for (size_t i = 0, n = nums.size(); i + 1 < n; ++i) {
            if (nums[i] > nums[i + 1]) {
                if (modified)
                    return false;
                // modifying strategy:
                // modify [i] to the least possible value,
                // if that's not possible, modify [i+1] to the least possible value.
                if (i == 0) {
                    nums[i] = -1e5;
                } else if (nums[i - 1] <= nums[i + 1]) {
                    nums[i] = nums[i - 1];
                } else {
                    nums[i + 1] = nums[i];
                }
                modified = true;
            }
        }
        
        return true;
    }
};
